flow/Deque.h

```diff
@@ -168,10 +168,11 @@ class Deque {
 		if (newSize > max_size()) throw std::bad_alloc();
 		//printf("Growing to %lld (%u-%u mask %u)\n", (long long)newSize, begin, end, mask);
 		T* newArr = (T*)aligned_alloc(std::max(__alignof(T), sizeof(void*)),
-		                              newSize * sizeof(T)); // SOMEDAY: FastAllocator, exception safety
+		                              newSize * sizeof(T)); // SOMEDAY: FastAllocator
 		ASSERT(newArr != nullptr);
 		for (int i = begin; i != end; i++) {
-			new (&newArr[i - begin]) T(std::move(arr[i&mask]));
+			new (&newArr[i - begin]) T(std::move_if_noexcept(arr[i & mask]));
```

Comment:
I _think_ this expression can still throw. I think what this achieves is that either this won't throw or the original array won't be modified.

If anything throws here then I think we leak arr, and we can still leave things in an invalid state where e.g. front() returns a reference to a destroyed element.

Can we additionally static assert `std::is_nothrow_move_constructible_v<T> || std::is_nothrow_copy_constructible_v<T>` ? Then I think nothing in this loop should throw

Reply:
This assertion fails for many types `T`. Instead, to get around this issue, I moved the destructor calls into a separate loop, so that if a copy constructor throws, the `Deque` is still in a valid state, and `~Deque` will clean up `arr` when necessary, without the destructors being called twice.